File: python/solutions/day_01.py
from models.aoc_solution import AOCSolution


class Day01(AOCSolution):
    EXPECTED = {
        "part_one": {"sample": 3, "data": 1023},
        "part_two": {"sample": 6, "data": 5899},
    }

    @property
    def parsed_data(self) -> list[tuple[int, int]]:
        """Return tuple of direction and clicks"""
        return [(-1 if row[0] == "L" else 1, int(row[1:])) for row in self.data.splitlines()]

    def part_one(self) -> int:
        """How many times the dial lands on 0 after turning"""
        position = 50
        zeroes = 0
        for direction, clicks in self.parsed_data:
            position += clicks * direction
            position %= 100
            zeroes += position == 0
        return zeroes

    def part_two(self) -> int:
        """How many times the dial passes through 0 during turning"""
        position = 50
        zeroes = 0
        for direction, clicks in self.parsed_data:
            for _ in range(clicks):
                position += direction
                position %= 100
                zeroes += position == 0
        return zeroes


if __name__ == "__main__":
    Day01().run()

File: typescript/solutions/day_01.ts
const parse = (input: string) => {
  return input
    .split("\n")
    .map((row) => [row[0] === "L" ? -1 : 1, parseInt(row.slice(1))]);
};

export const part1 = (input: string) => {
  let position = 50;
  let zeroes = 0;
  parse(input).map(([direction, clicks]) => {
    position += clicks * direction;
    position %= 100;
    zeroes += position === 0 ? 1 : 0;
  });
  return zeroes;
};

export const part2 = (input: string) => {
  let position = 50;
  let zeroes = 0;
  parse(input).map(([direction, clicks]) => {
    const clicks_to_next_zero = (10000 - direction * position) % 100;
    zeroes += - Math.floor((clicks_to_next_zero - clicks) / 100);
    position += clicks * direction;
    position %= 100;
  });
  return zeroes;
};
